### ISSUE TITLE
User-defined actions on built-in buttons

### ISSUE THREAD

**Issue by DmitrySharabin** [mavoweb/mavo#506]
In Mavo there are built-in buttons (e.g. `Add`, `Edit` etc.) that perform predefined actions. Is there any chance we could add user-defined actions to these buttons that would be performed **before** the default Mavo actions? 
I have a case that needs such behavior: https://codepen.io/dsharabin/pen/vPgvNm?editors=1000
Or there is a workaround for such cases?

**Comment by DmitrySharabin** [comment]
Definitely, there is a workaround: `<button mv-action="<extra actions>, add(item)">Add item</button>`.
But to my mind, it's a bit hacky: in edit mode, we would have two `Add item` buttons by default, and we have to make some extra styling to hide the buttons when we don't need them. So, we have to recreate basic Mavo functionality. Or I am wrong, am I?

**Comment by DmitrySharabin** [comment]
It might be we would need to perform an extra action not *before* the default one, but *after* it. So, to my mind, it would be useful to have something like `mv-action-before` and `mv-action-after` for custom buttons like `Add item`.

**Comment by DmitrySharabin** [comment]
Or much simpler: just redefine the default action in the `mv-action` attribute if it would have priority over the action generated by Mavo. :)

Sorry for the flood! :)

**Comment by karger** [comment]
this starts to remind me of work from the meta-object protocol.   when 
you create subclasses, there are a variety of ways you might want to let 
the methods of the subclass affect the methods of the superclass.  as 
you observe, the subclass might want to do some stuff before the 
superclass does its thing, or after, or completely replace.  The 
metaobject protocol defined keywords :before :after and :around to 
specify behavior.  
http://www.lispworks.com/documentation/HyperSpec/Body/07_ffb.htm

------ Original Message ------

**Comment by LeaVerou** [comment]
So right now we can have custom add buttons by using an `mv-add` class (or `mv-add-propertyname`). So this could be as simple as not adding an action to that button if one exists already.
Same thing could work for the item bar, now that most of it is expressed via actions behind the hood. Though some of them are more complex. 

> Sorry for the flood! :)

Please don't apologize, we love it!

**Comment by DmitrySharabin** [comment]
> this starts to remind me of work from the meta-object protocol

That really does! :)

Thank you for the link. I didn't know that concept.

> Please don't apologize, we love it!

Thank you!
